.circleci/config.yml

```diff
@@ -0,0 +1,126 @@
+            >> $BASH_ENV echo 'export PATH="$HOME/.daml/bin:$PATH"'
+  install_yq:
+    description: "Install yq from binary"
+    steps:
+      - run:
+          name: Install yq
+          command: |
+            mkdir -p $HOME/yq
+            curl -L https://github.com/mikefarah/yq/releases/download/3.4.1/yq_linux_amd64 -o $HOME/yq/yq &&\
+              chmod +x $HOME/yq/yq &&\
+              echo 'export PATH=$HOME/yq/yq:$PATH' >> $BASH_ENV
+  install_gh:
+    description: "Install GitHub Cli"
+    steps:
+      - run:
+          name: Install gh
+          command: |
+            curl -fsSL https://cli.github.com/packages/githubcli-archive-keyring.gpg | sudo dd of=/usr/share/keyrings/githubcli-archive-keyring.gpg
+            echo "deb [arch=$(dpkg --print-architecture) signed-by=/usr/share/keyrings/githubcli-archive-keyring.gpg] https://cli.github.com/packages stable main" | sudo tee /etc/apt/sources.list.d/github-cli.list > /dev/null
+            sudo apt update
+            sudo apt install gh
+  set_sdk_version:
+    description: "Set Daml SDK version environment version"
+    steps:
+      - run:
+          name: Set Daml SDK version
+          command: echo 'export SDK_VERSION=$(yq e '.sdk-version' daml.yaml)' >> $BASH_ENV
+  run_release:
+    description: "Run Release"
+    steps:
+      - run:
+          name: "Get target release version"
+          command: echo 'export VERSION=$(yq e '.version' daml.yaml)' >> $BASH_ENV
+      - run:
+          name: "Validate target release version"
+          command: |
+            if [ `git tag | grep ${VERSION} | wc -l` -eq 0 ]; then
+              echo "New tag detected ${VERSION}. Releasing..."
+            else
+              echo "Tag ${VERSION} already exists; Skipping Release."
+              circleci-agent step halt
+            fi
+      - run:
+          name: "Tag and push to GitHub"
+          command: gh release create v${VERSION} .daml/dist/daml-ctl-${VERSION}.dar --generate-notes
+      - store_artifacts:
+          path: .daml/dist/daml-ctl-${VERSION}.dar
+          destination: daml-ctl-dar-${VERSION}
+
+jobs:
+  build:
+    executor: daml-docker-executor
+    steps:
+      - checkout
+      - install_yq
+      - restore_cache:
+          keys:
+            - daml-ctl-{{ checksum "daml.yaml" }}
+      - set_sdk_version
+      - install_sdk
+      - run:
+          name: Daml build
+          command: daml build
+      - run:
+          name: Daml test
+          command: daml test
+      - save_cache:
+          paths:
+            - .daml
+          key: daml-ctl-{{ checksum "daml.yaml" }}
+  release:
+    executor: daml-docker-executor
+    steps:
+      - checkout
+      - install_yq
+      - install_gh
+      - restore_cache:
+          keys:
+            - daml-ctl-{{ checksum "daml.yaml" }}
+      - run_release
+
+workflows:
+  version: 2
+  daml_ctl_build:
+    jobs:
+      - build:
+          filters:
+            branches:
+              ignore:
```

Comment:
Why ignore master? Is it because you do it in `daml_ctl_release`? Isn't it more logical to have the inverse (i.e. release only on master?)

Reply:
btw should we rename master -> main to be consistent across our repos?

Reply:
> Why ignore master? Is it because you do it in daml_ctl_release? Isn't it more logical to have the inverse (i.e. release only on master?)

For this workflow `daml_ctl_build`, we only run the `build` job - this workflow is intended for feature branches (or any branch thats not master) so we know the code builds and the tests passes (if there are any... cough cough) before merging to master (or any other branch). 

The release workflow is `daml_ctl_release` which only runs on master branch.

> btw should we rename master -> main to be consistent across our repos?

We can rename it.

Reply:
> so we know the code builds and the tests passes

But that doesn't guarantee that it will also pass on the master branch (think back to spider - this is only true if you've rebased/merged from master and nobody has committed in the meantime). So we should also test before release on master.

Reply:
Just for clarity: I'm saying we should be building and testing _both_ on master and on feature branches.

Reply:
There is no guarantee that any given branch which when merged to master will also build successfully after merging to master. 

Thats why the `daml_ctl_release` workflow runs the `build` job (the same job thats ran all other branches) on the master branch **before** running the release job on every merge to master (see `requires` in the `release` configuration below - if the `build` job fails, the `release` job will not run) :

```
workflows:
  ...
  daml_ctl_release:
    jobs:
      - build:
          filters:
            branches:
              only:
                - master
      - release:
          context: github-fin-eng-context
          requires:
            - build
          filters:
            branches:
              only:
                - master
```

Reply:
> Just for clarity: I'm saying we should be building and testing both on master and on feature branches.

We are testing on both - you're not understanding the configuration correctly. Happy to do a call to explain.